### ISSUE TITLE
OnClik in PanelHolder.Tab()

### ISSUE THREAD

**Issue by DoctorovEV** [google/gxui#102]
Ben, good health ! 
Strange behavior: if you click on the Tab quickly ceases to be called a handler. Try different options and see the wrong result.

package main

import (
 
    "github.com/google/gxui"
    "github.com/google/gxui/drivers/gl"
    "github.com/google/gxui/themes/dark"
    "strconv"
    "fmt"
)

func appMain(driver gxui.Driver) {
    theme := dark.CreateTheme(driver)

    holder := theme.CreatePanelHolder()
    for i:=0; i<15 ; i++ {
        image := theme.CreateImage()
        holder.AddPanel(image, "New "+strconv.Itoa(i))
        a := i
        holder.Tab(i).OnClick(func (gxui.MouseEvent) {fmt.Println(a)})
    }
    window := theme.CreateWindow(800, 600, "test the txt box")
    window.AddChild(holder)
    window.OnClose(driver.Terminate)

}
func main() {
    gl.StartDriver(appMain)
}

**Comment by ben-clayton** [comment]
That is very strange indeed. I'll investigate...

**Comment by ben-clayton** [comment]
So the issue is when the tab is dragged - the dragged tab control is destroyed and a new one is created in its place, losing the event subscriptions.
I'm thinking the [PanelHolder](https://github.com/google/gxui/blob/master/panel_holder.go#L16) interface is broken. [PanelHolder.Tab()](https://github.com/google/gxui/blob/master/panel_holder.go#L16) should probably be removed and PanelHolder should adopt the regular [Container](https://github.com/google/gxui/blob/master/container.go) interface with:
```go
Name(Control) string
SetName(Control, string)
```
methods.

Why do you want to know if a tab was clicked? Aren't you really interested in knowing whether the panel is visible? If so, you can always listen to [Control.OnAttach](https://github.com/google/gxui/blob/master/control.go#L44) and [Control.OnDetach](https://github.com/google/gxui/blob/master/control.go#L45) on the panel. This approach will also work if the tab is dragged to another panel holder, unlike listening to [OnClick](https://github.com/google/gxui/blob/master/control.go#L48) or using [PanelHolder.SelectedPanel](https://github.com/google/gxui/issues/101).

**Comment by ben-clayton** [comment]
Proposed new interface:

```go
type PanelHolder interface {
	Control
	Container

	// Select will make the specified Control the selected panel.
	// Select will panic if panel is not a child of this PanelHolder.
	Select(panel Control)

	// Selected returns the currently selected panel, or nil if there are no 
	// panels in this PanelHolder.
	Selected() Control

	// Name returns the name of the specified child panel.
	// Name will panic if panel is not a child of this PanelHolder.
	Name(panel Control) string

	// SetName sets the name of the specified child panel, updating the tab text.
	// SetName will panic if panel is not a child of this PanelHolder.
	SetName(panel Control, name string)
}
```

**Comment by DoctorovEV** [comment]
Yes, good, Control.OnAttach and Control.OnDetach and Name, SetName solve the most problems.

What is the general event is used when replacing the panel to configure the environment ? For example, to clear the current action.

**Comment by ben-clayton** [comment]
> Although if you store a pointer to the Control name do not need to...

Care to elaborate?

> What is the general event is used when replacing the panel to configure the environment ? For example, to clear the current action.

I'm afraid I'm not following again. What action? Are you asking how do you remove all the panels?

**Comment by DoctorovEV** [comment]
I need the MAIN EVENT(for ex: OnSwitch() EventSubscription) when switching between panels. Nothing more.

**Comment by ben-clayton** [comment]
I'd like to know why you need this.
Adding events to fire when a panel is made visible in a particular PanelHolder sounds like a bug waiting to happen. Panels can be dragged between PanelHolders - if you want to know whether a panel is visible then use OnAttach and OnDetach. If there's some other reason you'd like to be told when a panel is made active in a single PanelHolder, then I'd like to know what the usage case is before implementing it.

**Comment by DoctorovEV** [comment]
Control.OnAttach and Control.OnDetach must be specified in each panels. Is this not excessive if you want to produce one and the same operation when changing tabs ?
a simple example: a variable dragged = 1. When changing tabs must dragged = 0. How it can be easy to implement?

**Comment by ben-clayton** [comment]
> Control.OnAttach and Control.OnDetach must be specified in each panels. Is this not excessive if you want to produce one and the same operation when changing tabs ?

I don't see it is as excessive as it's the correct thing to do, and I don't think it has a common usage case. As I keep on saying, I believe that having a PanelSelected event will just encourage people to do the wrong thing. If you can give me a reason why knowing what panel is active for a single PanelHolder I'll consider adding it. 

> a simple example: a variable dragged = 1. When changing tabs must dragged = 0. How it can be easy to implement?

What's being dragged? For what duration is dragged == 0?

I know the language barrier is making this hard, but please describe what you are trying to do, and hopefully we can come to some sort of compromise.

**Comment by DoctorovEV** [comment]
Thank Ben, if you believe that having a PanelSelected event will just encourage people to do the wrong thing, then I would agree with your professional opinion. I think more on this topic.

Using OnAttach and OnDetach everything works fine.

**Comment by ben-clayton** [comment]
I still need to implement the interface as discussed.
